more safe includes on application start

ktomk opened the pull request:
the include operations were not isolated then glob'ed registration.php and
cli_commands.php are automatically included.

this commit separates the globl-list from the code to include and the
inclusion is done in a separated scope.

Vinai commented on app/etc/NonComposerComponentRegistration.php:
```diff
@@ -4,19 +4,38 @@
-    }
-    foreach ($files as $file) {
-        include $file;
+namespace Magento\NonComposerComponentRegitsration;
```
Typo: `Regitsration` should be `Registration`

Vinai commented:
Interesting, the change seems to cause [MemoryUsageTest](https://github.com/magento/magento2/blob/develop/dev/tests/integration/testsuite/Magento/MemoryUsageTest.php#L33) to fail. On my local dev platforms the test is a little flaky, so lets see if the next build happens to pass.

Vinai commented on app/etc/NonComposerComponentRegistration.php:
```diff
@@ -4,19 +4,38 @@
-    }
-    foreach ($files as $file) {
-        include $file;
+namespace Magento\NonComposerComponentRegitstration;
```
Typo in namespace `Regitstration` should be `Registration`

ktomk commented:
Just corrected the typo.

For the test i have to take a look. Behaviour should not have changed though.

Vinai commented:
Sorry, typo still is present (but changed).

ktomk committed:
more safe includes on application start

the include operations were not isolated then glob'ed registration.php and
cli_commands.php are automatically included.

this commit separates the globl-list from the code to include and the
inclusion is done in a separated scope.

ktomk commented:
Good morning :) Thx.

Vinai commented:
:coffee:

ktomk commented:
@Vinai Can you give me a link/hint how I do run that testsuite locally? Then I can better check for the memory leak test.

Vinai commented:
Since it is an integration test, you will first have to configure a testing db in `dev/tests/integration/etc/install-config-mysql.php` (copy from `.dist`).
Then to run the test do
`cd dev/tests/integration && ../../../vendor/bin/phpunit testsuite/Magento/MemoryUsageTest.php`

It will take a while because it installs into the test db in the background.
Edit: Oh, @davidalger wrote a [nice blog post](http://davidalger.com/development/magento/running-the-magento-2-test-suite/) on how to run the Magento 2 testsuites if you want more detail:

ktomk commented:
@Vinai Thx for the pointers, If I run the test isolated, it works on my end. If I run all tests the one in question doesn't fail as well but only this one 19 times (which looks like a test-suite-setup inconsistency to me):

```
19) Magento\Test\Integrity\Magento\Payment\MethodsTest::testPaymentMethod with data set #19 ('hosted_pro', 'Magento\\Paypal\\Model\\Hostedpro')
UnexpectedValueException: RecursiveDirectoryIterator::__construct(/home/user/PhpstormProjects/magento2/dev/tests/integration/testsuite/Magento/Test/Integrity/Magento/Payment/../../../../../Magento/Framework/Filesystem/_files/generated): failed to open dir: Permission denied

/home/user/PhpstormProjects/magento2/dev/tests/integration/testsuite/Magento/Test/Integrity/Magento/Payment/MethodsTest.php:139
/home/user/PhpstormProjects/magento2/dev/tests/integration/testsuite/Magento/Test/Integrity/Magento/Payment/MethodsTest.php:29
```

The permissions of that folder:

```
d-wxrw--wt 3 user user 4096 Jan 26 13:01 generated
```

no read rights for user and world (?!)

Full listing:

```
$ ls -al /home/user/PhpstormProjects/magento2/dev/tests/integration/testsuite/Magento/Test/Integrity/Magento/Payment/../../../../../Magento/Framework/Filesystem/_files
total 28
drwxrwxr-x 4 user user 4096 Jan 26 13:01 .
drwxrwxr-x 6 user user 4096 Jan 12 18:42 ..
-rw-rw-r-- 1 user user  197 Jan 12 18:42 ClassToFind.php
-rw-rw-r-- 1 user user   38 Jan 12 18:42 data.csv
drwxrwxr-x 3 user user 4096 Jan 12 18:42 foo
d-wxrw--wt 3 user user 4096 Jan 26 13:01 generated
-rw-rw-r-- 1 user user   15 Jan 12 18:42 popup.csv
```

My PHP Version:

```
$ php -v
PHP 5.6.17-3+deb.sury.org~trusty+1 (cli) 
Copyright (c) 1997-2015 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Technologies
    with Zend OPcache v7.0.6-dev, Copyright (c) 1999-2015, by Zend Technologies
```

And yes, it takes some time :)

ktomk committed:
rename NonComposerComponentRegistration to registration

- rename to registration

- rename the glob pattern-list-file accordingly

- streamline with autoload.php as it is on the same level, that is, effec-
  tively move the file one directory up.

- require file with list of glob patterns (squash)

buskamuza commented on app/registration.php:
```diff
@@ -0,0 +1,43 @@
+ * @return void
+ */
+function file() {
+    include func_get_arg(0);
```
Why you don't use explicit argument here?

ktomk replied:
To not use a variable.

buskamuza replied:
What's bad with using a variable?

ktomk replied:
I'd say bad/good is not the level that this is about. So technically I don't understand your question. It's a preference I have for this situation for the level of isolation. As I wrote, a variable does work as well and given the code verbatim, even if it (the parameter variable) is manipulated by the include it should have no consequences.

buskamuza commented on app/registration.php:
```diff
@@ -0,0 +1,43 @@
+        if ($files === false) {
+            throw new RuntimeException("glob(): error with '$baseDir$globPattern'");
+        }
+        array_map(__NAMESPACE__ . '\file', $files);
```
Using string name for a function makes its usages invisible. Is there any benefit this such implementation vs. iteration through the array and calling the function?

davidalger replied:
I would suggest changing this line to simple this. It accomplishes the objective of variable scope isolation, as well as eliminates the string reference to a callback function.

``` php
array_map(function ($file) { require_once $file; }, $files);
```

ktomk replied:
I was experimenting back and forth because of the problems in the build IIRC (integration testsuite). Usage should be local anyway, so usage might not be such a problem (and IIRC Phpstorm understands if a string represents a callback). In any way, it should be that variant taken that works best with the test-suite. I have no specific preference but _not_ creating a function object on the fly (as suggested by @davidalger).

davidalger commented on app/registration.php:
```diff
@@ -0,0 +1,43 @@
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\NonComposerComponentRegistration;
```
This should match the file name ;)

ktomk replied:
Actually not as it's not part of autoloading (no class). The namespace is just bulky to not be conflated with anything useful ;)

buskamuza replied:
Taking into account another @davidalger 's suggestion with anonymous function, maybe we can use anonymous function for the whole registration logic as well? So we should not care about names at all.

ktomk replied:
You just should not introduce anonymous functions for that.

mazhalai commented:
@ktomk integration tests are failing. Please fix and also merge with latest develop

mazhalai commented:
Internal ticket MAGETWO-49472

buskamuza commented on app/etc/registration_globlist.php:
```diff
@@ -0,0 +1,13 @@
+ * registration.php to generate a list of includes.
+ */
+return [
+    'code/*/*/cli_commands.php',
```
I'd suggest to use root of Magento as a base path (so the paths would start from 'app/code' instead of 'code').
The goal of such change is to make more intuitive for a future reader. I would assume that the reader, when opens the file, would guess from 2 options:
1. Either the files are relative to current folder (`app/etc`)
2. or the files are relative to the root folder

It's difficult to guess that they are relative to `app` because there is no context of `app` here.

ktomk replied:
I can't say if a different path makes it more intuitive or not. I give the benefit of the doubt, for my change it was only to streamline based on common base-path, so a technical decision. I found it overall hard(er) to grasp and I don't know if the path was part of it. Perhaps a well written comment would do better than a code / data change. I can also imagine that this might change in the future so I would not suggest to introduce hard rules here.

ktomk commented:
@mazhalai: Last time I checked the test-fail was not specific to the changes but in the test-suite. Has that changed? For updating, I have the feeling it's not yet clear what's preferred for the code.

mazhalai commented:
@ktomk You will need to investigate, since builds on develop branch are passing https://travis-ci.org/magento/magento2/builds/110949135

ktomk commented:
@mazhalai; No investigation necessary, rebased changes pass green. looks like problem was in base branch and even if it still is, for the moment it won't matter.

vrann commented:
@ktomk can you please update your branch with the latest mainline?

ktomk commented:
@vrann: rebased on top of develop 7c3dc9af160ad3fb45ab9fe1553e26b50ba28d0e.

vrann commented:
@ktomk thank you.

Tried to understand the purpose of the pull request. Can you please list out what advantages does such approach have vs what we already have in the codebase?

ktomk commented:
@vrann: Sure, it is also perhaps better to squash both commits together so it can be more easily reviewed:

There is a code change to clarify the wording and to separate the data from the actual code. This also helped to reduce dirname() calls to a great extend.

The other part is that inclusions are not any longer in the global scope (variables). This is good against unwanted side-effects. That is merely the main point, reduction of side-effects.

So in list form:

* Reduce `dirname()` calls.

* Reduce `__DIR__` usage

* Add another include (for the data)

* Add namespace (just for completeness what changed)

* Remove global scope while including.(main feature I'd say)

* Alignment of filename for the type of registration (`registration.php`) (side-note: I personally found it once a bit misleading when I read non.-composer as I used composer and this registration was effective regardless being non-composer)

okorshenko commented:
Hi @ktomk could you please resolve merge conflicts or allow change the code for Maintainers (see checkbox on the right). Thank you

ktomk commented:
checked.

okorshenko commented:
Thank you

Commit:
Merged latest changes from magento/magento2 : develop

Ctucker9233 commented:
@magento-team Can this be backported to 2.2?
